testng-core/src/test/kotlin/test/hook/HookableTest.kt

```diff
@@ -0,0 +1,285 @@
+package test.hook
+
+import org.assertj.core.api.Assertions.assertThat
+
+import java.lang.reflect.Method
+import java.util.UUID
+import java.util.function.Consumer
+import org.assertj.core.api.SoftAssertions
+import org.testng.*
+import org.testng.annotations.DataProvider
+import org.testng.annotations.Test
+import test.SimpleBaseTest
+import test.TestData
+import test.hook.samples.ConfigurableFailureSample
+import test.hook.samples.ConfigurableFailureWithStatusAlteredSample
+import test.hook.samples.ConfigurableSuccessSample
+import test.hook.samples.ConfigurableSuccessWithListenerSample
+import test.hook.samples.HookFailureSample
+import test.hook.samples.HookFailureWithStatusAlteredSample
+import test.hook.samples.HookSuccessDynamicParametersSample
+import test.hook.samples.HookSuccessSample
+import test.hook.samples.HookSuccessTimeoutSample
+import test.hook.samples.HookSuccessTimeoutWithDataProviderSample
+import test.hook.samples.HookSuccessWithListenerSample
+import test.hook.samples.issue2251.TestCaseSample
+import test.hook.samples.issue2257.TestClassSample
+import kotlin.reflect.KClass
+
+const val HOOK_INVOKED_ATTRIBUTE = "hook"
+const val HOOK_METHOD_INVOKED_ATTRIBUTE = "hookMethod"
+const val HOOK_METHOD_PARAMS_ATTRIBUTE = "hookParamAttribute"
+
+class HookableTest : SimpleBaseTest() {
+
+    @Test(dataProvider = "getTestClasses")
+    fun hookSuccess(clazz: KClass<*>, flow: String, assertAttributes: Boolean) {
+        Reporter.log("Running scenario $flow", true)
+        val listener = TestResultsCollector()
+        create(clazz.java).run {
+            addListener(listener)
+            run()
+        }
+        assertThat(listener.getPassedMethodNames()).contains("verify")
+        if (!assertAttributes) {
+            return
+        }
+        val assertions = SoftAssertions()
+        listener.getPassed().forEach {
+            assertions.assertThat(it.getAttribute(HOOK_INVOKED_ATTRIBUTE)).isNotNull
+            assertions.assertThat(it.getAttribute(HOOK_METHOD_INVOKED_ATTRIBUTE)).isNotNull
+            if (it.method.isDataDriven) {
```

Comment:
Yes it is, but the same data provider is facilitating the tests. This is the only diff between this and a data provider powered test. Hence am clubbing them instead of just duplicating the test code.